src/network.rs

```diff
@@ -0,0 +1,75 @@
+// Written by the rust-bitcoin developers.
+// SPDX-License-Identifier: MIT
+
+//! A type representing the current Bitcoin networks.
+//!
+//! Used for getting access to the human-readable part of a bech32 encoded address as defined by
+//! [BIP-173].
+//!
+//! Please note this does not include signet because testnet and signet use the same HRP ("tb").
+//!
+//! [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+
+/// A subset of the Bitcoin networks, which use a unique HRP.
+#[derive(Copy, PartialEq, Eq, Clone, Hash, Debug)]
+#[non_exhaustive]
+pub enum Network {
+    /// Mainnet Bitcoin - "bc".
+    Bitcoin,
+    /// Bitcoin's testnet network - "tb".
+    Testnet,
+    /// Bitcoin's regtest network - "bcrt".
+    Regtest,
+}
+
+impl Network {
+    /// Converts a [`Network`] to a string used as the human-readable part of an address encoded
+    /// using bech32 and the [segwit address format] as defined in [BIP-173].
+    ///
+    /// Returns the human-readable part for this network.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use bech32::Network;
+    /// assert_eq!(Network::Bitcoin.to_hrp(), "bc");
+    /// assert_eq!(Network::Testnet.to_hrp(), "tb");
+    /// assert_eq!(Network::Regtest.to_hrp(), "bcrt");
+    /// ```
+    ///
+    /// [segwit address format]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Segwit_address_format>
+    pub const fn to_hrp(&self) -> &'static str {
```

Comment:
Why not return an instance of the human-readable part type instead of a string slice?

Reply:
I'm having a play with introducing the `Hrp` type as you suggested previously, cheers.

Reply:
Aight, I hacked up a new version of this PR that adds an `Hrp` type, and also a `KnownHrp` type for segwit specific code. Could be mistakes because it was a fair bit of work and I should really go over it on another day before pushing but I'm pretty psyched for this so if you'll excuse me I'll push right away.